### ISSUE TITLE
Silent chunking failure for dimensions not in name_maps

### ISSUE THREAD

**Issue by angus-g** [OceanParcels/parcels#768]
The automatic chunking mechanism (e.g. passing a tuple of chunk sizes) determines dimension names for each variable independently, which has no problems assuming automatic chunking is supported.

For files with dimensions that aren't in `parcels.field.NetcdfFileBuffer._name_maps`, we have to specify chunks as a dictionary. However, because of staggering, different variables have different dimension names. This chunking dictionary is passed directly to xarray, which throws an error if extra dimensions are present.

## Example

For example, we have B-grid data which is all chunked in the same way:
```python
chunks = {
  "st_ocean": 7, "sw_ocean": 7,
  "xt_ocean": 400, "xu_ocean": 400,
  "yt_ocean": 300, "yu_ocean": 300,
}
```

This gives the following error:

```
ValueError: some chunks keys are not dimensions on this object: {'yt_ocean', 'xt_ocean', 'sw_ocean'}
```

At this point, I reduced the chunks to only the subset of dimensions that appear on the first field that's processed:
```python
chunks = {
  "st_ocean": 7,
  "xu_ocean": 400,
  "yu_ocean": 300,
}
```

However, because the dimension names are still not expected, the chunksize/chunkmap mechanism silently breaks. The relevant part of the traceback is:

```
  File "/nix/store/9racrka4shwcahxrbzk14w6lpzbp3nrs-python3.7-parcels/lib/python3.7/site-packages/parcels/field.py", line 1976, in data_access
    data = data.rechunk(self.chunk_mapping)
```

Which leads to a division by zero in dask: `self.chunk_mapping = {0: 0, 1: 0}`.

## Thoughts
One way to get around this would be just to add the extra dimension names to the name map, but this doesn't scale to the general case. It does seem that the whole chunking mechanism is a bit complicated and magic.

**Comment by CKehl** [comment]
well, Angus, so this problem is connected to the connection between netcdf file and the Grid and the Field that are using this file. So, I would opt e.g. for the first example
```
chunks = {
  "st_ocean": 7, "sw_ocean": 7,
  "xt_ocean": 400, "xu_ocean": 400,
  "yt_ocean": 300, "yu_ocean": 300,
}
```
because the function `_get_initial_chunk_dictionary` in the NetcdfFileBuffer will try then to map those provided dimensions to the largest subset of parameters for a variable (which by assumption is the actual data variable) and apply those numbers to it (line 1808 to 1839 in field.py).

the `_name_map` tries to provide a link between the supplied chunk dimensions in netcdf and the actual dimensions in parcels. If you have a good alternative approach how to establish this connection, then I'd be happy to hear about it. As you said: the problem is that those dimensions in any netcdf can be named arbitrarily, so mapping those netcdf dimensions to 'time', 'depth', 'lat' and 'lon' in a generally-applicable manner is undefined. The dictionary solution was just the easiest and quickest to do that.

Your proposal with the extra dimensions to the `dimensions` dictionary is quite good, and I will think about it.
The other point is that this chunking is also connected to the grid: 2 fields can share the same grid in terms of dimension names and indices, but actually have different chunking sizes or access different netcdf dimensions - which actually makes then being on separate grids. That conflicts e.g. for NEMO data and C-grid data in general, where at the very least U and V need to be on the same grid. How this can be done is still not completely ironed out.

I agree that the chunking is a bit complicated (though not magic - it just requires dedicated memory management, and that is not really the strength of Python), but it's worth doing it because the size and resolution of the grids are ever increasing, and dynamic data loading via field chunking is per-se the best option to facilitate computations on virtually infinitely-large grids.

The whole thing will also be not-applicable when at some point dealing with unstructured grids, but that's gonna be a point to address when getting there.

From an engineering perspective and code refinement, we could brain-storm already about externalising the `FileBuffer` mechanism from the field, and provide a `DynamicFileAccessor` (w. chunking), a `BufferedFileAccessor` (w. just deferred loading but no chunking) and a `FixedFileAccessor` (with neither - loading the full field via xarray as numpy in one go). That will then also have implications and require restructuring for fields, grids and their sets.

Also, this whole part assumes that we have a `FileBuffer` to access. @erikvansebille and me also have some talks to researchers to interface on-the-fly OCM sources in an on-line computation - obviously neither working with dynamic loading or any of those solutions discussed above. That means that all this will anyway require a bit of re-structuring in the mid-to-long term future.

**Comment by stellema** [comment]
I've been running into this issue using the same 3D OFAM b-grid as the example. In the meantime, is setting `field_chunksize="auto"` in `FieldSet.from_b_grid_dataset` the only option for chunking this type of grid? 
As in the example below, specifying` field_chunksize="auto"`, throws a warning and ends up using the grid's master chunksize. 
```
u = ['ocean_u_1981_01.nc']  # Lat/lon subset of 0.1 deg grid with daily output.
v = ['ocean_v_1981_01.nc']
w = ['ocean_w_1981_01.nc']
files = {'U': {'depth': w[0], 'lat': u[0], 'lon': u[0], 'data': u},
         'V': {'depth': w[0], 'lat': u[0], 'lon': u[0], 'data': v},
         'W': {'depth': w[0], 'lat': u[0], 'lon': u[0], 'data': w}}
variables = {'U': 'u', 'V': 'v', 'W': 'w'}
dimensions = {'time': 'Time', 'depth': 'sw_ocean', 'lat': 'yu_ocean', 'lon': 'xu_ocean'}
fieldset = FieldSet.from_b_grid_dataset(files, variables, dimensions, mesh='spherical',
                                        field_chunksize='auto')
```
output:
``` 
WARNING: Trying to initialize a shared grid with different chunking sizes - action prohibited. 
Replacing requested field_chunksize with grid's master chunksize.
```
However, I'm not sure if any chunking is being done and is there is anyway to improve it?
``` 
fieldset.computeTimeChunk(0, 1)
print(fieldset.U.__dict__)
```
output:
```
{'name': 'U', 'filebuffername': 'u', 
'data': dask.array<where, shape=(3, 51, 300, 1750), dtype=float32, chunksize=(1, 51, 300, 731), chunktype=numpy.ndarray>, 
'grid': <parcels.grid.RectilinearZGrid object at 0x000001DD8F77D5F8>, 
'igrid': 0, 'lon': array([removed]), 'lon': array([removed]),
'fieldtype': 'U', 'units': <parcels.tools.converters.GeographicPolar object at 0x000001DD980F39B0>, 
'timestamps': None, 'interp_method': 'bgrid_velocity', 
'fieldset': <parcels.fieldset.FieldSet object at 0x000001DD98F9E748>, 
'allow_time_extrapolation': False, 'time_periodic': False, 'vmin': None, 'vmax': None, '_scaling_factor': None, 
'dimensions': {'time': 'Time', 'depth': 'sw_ocean', 'lat': 'yu_ocean', 'lon': 'xu_ocean'}, 
'indices': {'lon': range(0, 1750), 'lat': range(0, 300), 'depth': range(0, 51)}, 
'dataFiles': array([removed]), 
'netcdf_engine': 'netcdf4', 'loaded_time_indices': range(0, 3), 'creation_log': 'from_b_grid_dataset', 
'field_chunksize': 'auto', 'data_full_zdim': 51, 
'data_chunks': [None, None, None], 'c_data_chunks': [None, None, None], 
'nchunks': (3, 1, 1, 3),  'chunk_set': True, 
'filebuffers': [<parcels.field.NetcdfFileBuffer object at 0x000001DD9997B358>, <parcels.field.NetcdfFileBuffer object at 0x000001DD9998DEB8>, <parcels.field.NetcdfFileBuffer object at 0x000001DD999BDC18>]}
```

**Comment by CKehl** [comment]
Hi,

1) yes, those OFAM dimensions are not integrated right now in Parcels for chunking. A better way for that mapping needs to be thought of and implemented than the ´_name_map´, Perhaps I find some time to work on that soon.
2) I see that you really need the chunking (1750x300x51). The older-version alternative is still using indices, if that is an option for now.
3) You can define those chunk dimensions yourself in form of a dict. For that though, you need to know the name of the NetCDF dimensions;
4) It performs chunking - see your output printout, the ´'data'´ line.

We can chat a bit more on your case after my vacations. Cheers,

Christian

**Comment by stellema** [comment]
Thanks @CKehl, that would be great!

**Comment by erikvansebille** [comment]
I saw in some code that @hrsdawson sent me, that she used the following in her Parcels script, before creating the `FieldSet`:
```python
parcels.field.NetcdfFileBuffer._name_maps = {
    "lon": ["xu_ocean", "xt_ocean"],
    "lat": ["yu_ocean", "yt_ocean"],
    "depth": ["st_ocean", "sw_ocean"],
    "time": ["time"]}
```

@CKehl, would this be a possible solution to this name_map problem? If so, should we document this workaround and/or make this a prettier function?

**Comment by CKehl** [comment]
Hi,
well, for now and as long as this issue is open: yes, that's a work-around. That said, because it's so dirty, I'd not include it in the public documentation (personal intuition). Instead, it would be good to rather soon fix this in a suitable manner.
Cheers,
Christian

**Comment by CKehl** [comment]
Hey @stellema - please look at #853 and the [parcels_contributions repository](https://github.com/OceanParcels/parcels_contributions) for examples of how the name mapping with OFAM would work soon in the main branch. Also, feel free to comment on #853 if something is unclear or if you have further feedback on it. Cheers, Christian

**Comment by stellema** [comment]
Defining the dimensions in `name_maps` fixed my issue. Thank you @CKehl @erikvansebille! 
Cheers, Annette

**Comment by erikvansebille** [comment]
We've just discussed this with @ckehl, and the proposal is that we implement the following signature for specific Field chunksizes

`field_chunksize = {parcels_varname: {netcdf_dimname : (parcels_dimname, chunksize_as_int)}, ...}`
where `parcels_dimname` is one of (`time`, `depth`, `lat`, `lon`)

So e.g 
```python
field_chunksize = {'U': {'xu_ocean': ('lon', 128), 'yu_ocean': ('lat', 128), 'zt_ocean': ('depth', 3)},
                   'V': {'xv_ocean': ('lon', 128), 'yv_ocean': ('lat', 128), 'zt_ocean': ('depth', 3)},
                   'W': {'xt_ocean': ('lon', 128), 'yt_ocean': ('lat', 128), 'zw_ocean': ('depth', 3)}}
```

The idea is to implement this in #932

**Comment by CKehl** [comment]
actually, when rethinking our conversation and the code, the above definition is non-smart. better:

```
field_chunksize = {'U': {'lon': ('xu_ocean', 128), 'lat': ('yu_ocean', 128), 'depth': ('zt_ocean', 3)},
                   'V': {'lon': ('xv_ocean', 128), 'lat': ('yv_ocean', 128), 'depth': ('zt_ocean', 3)},
                   'W': {'lon': ('xt_ocean', 128), 'lat': ('yt_ocean', 128), 'depth': ('zw_ocean', 3)}}
```

**Comment by erikvansebille** [comment]
This is now fixed in #932
